scheduler/scheduler.go

```diff
@@ -1,122 +1,328 @@
+		queue: jobqueue.With(jobqueue.Options{
+			MaxConcurrency: c.MaxConcurrency,
+			MaxStackSize:   c.MaxQueueSize,
+			Timeout:        c.Timeout,
+		}),
+	}
 
-func (r *Registry) setStack(name string, s *Stack) {
-	r.stacks[name] = s
+	if measure {
+		if name == "" {
+			name = "unknown"
+		}
+
+		q.activeRequestsMetricsKey = fmt.Sprintf("lifo.%s.active", name)
+		q.queuedRequestsMetricsKey = fmt.Sprintf("lifo.%s.queued", name)
+		r.measure()
+	}
+
+	return q
 }
 
-// Do implements routing.PostProcessor and sets the stack for the scheduler filters.
-//
-// It preserves the existing stack when available.
-func (r *Registry) Do(routes []*routing.Route) []*routing.Route {
+func (r *Registry) initLIFOFilters(routes []*routing.Route) []*routing.Route {
 	rr := make([]*routing.Route, len(routes))
+	existingKeys := make(map[string]bool)
+	groups := make(map[string][]GroupedLIFOFilter)
+
 	for i, ri := range routes {
 		rr[i] = ri
 		for _, fi := range ri.Filters {
+			// TODO: warn on multiple lifos in the same route
+			if glf, ok := fi.Filter.(GroupedLIFOFilter); ok {
+				groupName := glf.Group()
+				groups[groupName] = append(groups[groupName], glf)
+				continue
+			}
+
 			lf, ok := fi.Filter.(LIFOFilter)
+			if !ok {
+				continue
+			}
+
+			var q *Queue
+			key := fmt.Sprintf("lifo::%s", ri.Id)
+			existingKeys[key] = true
+			c := lf.Config()
+			qi, ok := r.queues.Load(key)
 			if ok {
-				lf.SetKey(ri.Id)
-				key := lf.Key()
-				c := lf.Config(r)
-				s, ok := r.getStack(key)
-				if !ok {
-					s = newStack(c)
-					r.setStack(key, s)
-				} else if c != s.config { // UpdateDoc
-					s.config = c
-					r.setStack(key, s)
-				}
-
-				lf.SetStack(s)
+				q = qi.(*Queue)
+			}
+
+			if !ok {
+				q = r.newQueue(ri.Id, c)
+				r.queues.Store(key, q)
+			} else if q.config != c {
+				q.config = c
+				q.reconfigure()
+			}
+
+			lf.SetQueue(q)
+		}
+	}
+
+	for name, group := range groups {
+		var (
+			c           Config
+			foundConfig bool
+		)
+
+		for _, glf := range group {
+			if !glf.HasConfig() {
+				continue
+			}
+
+			if foundConfig && glf.Config() != c {
+				log.Warnf("Found mismatching configuration for the LIFO group: %s", name)
+				continue
 			}
+
+			c = glf.Config()
+			foundConfig = true
+		}
+
+		var q *Queue
+		key := fmt.Sprintf("group-lifo::%s", name)
+		existingKeys[key] = true
+		qi, ok := r.queues.Load(key)
+		if ok {
+			q = qi.(*Queue)
+		}
+
+		if !ok {
+			q = r.newQueue(name, c)
+			r.queues.Store(key, q)
+		} else if q.config != c {
+			q.config = c
+			q.reconfigure()
+		}
+
+		for _, glf := range group {
+			glf.SetQueue(q)
 		}
 	}
+
+	r.queues.Range(func(key, _ interface{}) bool {
+		if !existingKeys[key.(string)] {
+			r.queues.Delete(key)
+		}
+
+		return true
+	})
+
 	return rr
 }
 
-func (r *Registry) Close() {
-	for _, s := range r.stacks {
-		s.close()
+// Do implements routing.PostProcessor and sets the queue for the scheduler filters.
+//
+// It preserves the existing queue when available.
```

Comment:
Please mention side effect in the docs: Starts goroutine to create gauges, if `r.options.Metrics` is not nil and `measuring` is not true. (why do we need r.measuring? Can't we just use the nil check?)

Reply:
the `r.measuring` is used to avoid starting multiple goroutines and starting them lazily only when they're required for the first time. Since the post-processor can run only on a single goroutine for other reasons anyway, this approach seems OK here.

Reply:
ok